### PR TITLE
fix(editor): Remove invalid connections after node handles change

### PR DESCRIPTION
**Pull request by alexgrozav**
## Summary

https://github.com/user-attachments/assets/fd483573-ea7e-4ffd-beee-2e411df0242a

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/CAT-403/delete-connection-when-switching-to-node-that-doesnt-have-connection

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/editor-ui/src/utils/objectUtils.ts [hunk 1]
```diff
@@ -18,3 +18,48 @@ export const searchInObject = (obj: ObjectOrArray, searchString: string): boolea
 			? searchInObject(entry, searchString)
 			: entry?.toString().toLowerCase().includes(searchString.toLowerCase()),
 	);
+
+/**
+ * Efficiently and deeply compare two objects. Handles nested objects and arrays.
+ * Returns true if they are equal, false otherwise.
+ */
+export function deepCompare<T extends object>(
```
Comment (OlegIvaniv) [inline packages/editor-ui/src/utils/objectUtils.ts:26]: Do we need to implement custom method for this? 🤔 We're already using `isEqual` from `lodash-es` in our codebase which also does a deep comparison

File: packages/editor-ui/src/composables/useCanvasOperations.ts [hunk 2]
```diff
@@ -1230,11 +1231,85 @@ export function useCanvasOperations({ router }: { router: ReturnType<typeof useR
...
+						connection.data.source,
+						connection.data.target,
+					)
+				) {
+					void nextTick(() => deleteConnection(connection));
+				}
+			}
+		});
+	}
+
```
Comment (OlegIvaniv) [inline packages/editor-ui/src/composables/useCanvasOperations.ts:1307]: These two methods are pretty much identical so we could keep it DRY:
```suggestion
function revalidateNodeConnections(id: string, connectionType: 'input' | 'output') {
		const node = workflowsStore.getNodeById(id);
		const isInput = connectionType === 'input';
		if (!node) {
			return;
		}

		const nodeType = nodeTypesStore.getNodeType(node.type, node.typeVersion);
		if (!nodeType) {
			return;
		}

		const connections = mapLegacyConnectionsToCanvasConnections(
			workflowsStore.workflow.connections,
			workflowsStore.workflow.nodes,
		);

		connections.forEach((connection) => {
			const isRelevantConnection = isInput ? connection.target === id : connection.source === id;

			if (isRelevantConnection) {
				const otherNodeId = isInput ? connection.source : connection.target;

				const otherNode = workflowsStore.getNodeById(otherNodeId);
				if (!otherNode || !connection.data) {
					return;
				}

				const [firstNode, secondNode] = isInput ? [otherNode, node] : [node, otherNode];

				if (
					!isConnectionAllowed(
						firstNode,
						secondNode,
						connection.data.source,
						connection.data.target,
					)
				) {
					void nextTick(() => deleteConnection(connection));
				}
			}
		});
	}

	function revalidateNodeInputConnections(id: string) {
		return revalidateNodeConnections(id, 'input');
	}

	function revalidateNodeOutputConnections(id: string) {
		return revalidateNodeConnections(id, 'output');
	}
```

**Review by OlegIvaniv — COMMENTED** [review]
Works as expected just 2 comments that are up for discussion

File: packages/editor-ui/src/utils/objectUtils.ts (hunk 1 above)
Comment (alexgrozav) [reply packages/editor-ui/src/utils/objectUtils.ts]: I reinvented the wheel 😅... Thanks for pointing it out. Removed the helper.

File: packages/editor-ui/src/composables/useCanvasOperations.ts (hunk 2 above)
Comment (alexgrozav) [reply packages/editor-ui/src/composables/useCanvasOperations.ts]: Great idea. Applied it.

**Review by OlegIvaniv — APPROVED** [review]
Thanks, LGTM!

**Review by OlegIvaniv — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.73.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.73.0)

**Pull request merged**